docs/madr/decisions/042-tcp-grpc-probe.md

```diff
@@ -0,0 +1,246 @@
+
+#### Virtual Probes Listeners
+
+We generate a TCP proxy listener for each of gRPC and TCP probes, and forward the probe traffic transparently to the application. We also keep the existing HTTP Virtual Probe listener for HTTP probes.
+
+This keep the compatibility with the existing HTTP based Virtual Probe and data plane data model.
+
+#### Data Plane Data Model 
+
+Introduce two new fields onto the `Dataplane` resource type to store the Virtual Probe ports:
+
+```yaml
+  probes:
+    endpoints:
+    - inboundPath: /metrics
+      inboundPort: 8081
+      path: /8081/metrics
+    port: 19000
+  tcpProbes:
+  - port: 19001
+    applicationPort: 1433
+  grpcProbes:
+  - port: 19002
+    applicationPort: 2379
+    service: abcd
+```
+
+#### Positive Consequences
+
+- Ease of implementation
+
+- Keeping the compatibility with the existing HTTP based Virtual Probe and data plane data model
+
+#### Negative Consequences
+
+- The user experience of specifying virtual probes ports is more complex
+
+- More virtual listeners on sidecar
+
+## Other options
+
+### Merging all probes into one listener
+
+Introducing an intermediate layer to translate HTTP probes into TCP and gRPC probes (the translator), transform user specified probes (the user probe) into HTTP probes on the pod, and:
+
+1. If the user probe is HTTP based, requests are forwarded/redirected to application directly
+
+2. If the user probe is TCP or gRPC based, we'll still receive HTTP requests for these probes, since they are transformed to HTTP probes. We forward these requests to the translator and the translator is responsible for detecting the actual probe status from the application.
+
+The whole workflow can be shown in the following diagram:
+
+```
+  gRPC Probes                                               HTTP to gRPC translator --->  Application
+              ↘                                           ↗
+HTTP Probes --> HTTP Probes --->  Virtual Probe Listener ->  HTTP Probes  --->  Application
+              ↗                                           ↘
+   TCP Probes                                               HTTP to TCP translator  --->  Application
+```
+
+#### Positive Consequences
+
+- Don't need to introduce new ports for TCP and gRPC probes
+
+- Better user experience: don't need to introduce new annotations
+
+- Less virtual listeners on sidecar
+
+- Don't need to introduce new fields on the `Dataplane` resource type
+
+#### Negative Consequences
+
+- Requires an extra component (the translator) in `kuma-dp`, which increases the implementation complexity
+
+- Less intuitiveness, so harder for troubleshooting issues
+
+## Decision Drivers
+
+- Provide the best user experience by designing a simple and intuitive way of declaring virtual probes
+
+- Keep the compatibility with the existing HTTP based Virtual Probe and data plane data model
+
+## Decision Outcome
+
+Chosen option: (to be chosen)
```

Comment:
Two questions.

1) TCP probes
Do we even need to do anything special for TCP probes? Correct me if I'm wrong, but I think TCP probe just opens a connection and that's it. The connection can be mtlsed, because we don't send anything on it.

If we don't need to do anything about TCP then...
2) gRPC probe
gRPC is just HTTP under the hood. Do we even need to have any translator? Isn't it just a matter of providing proper HTTP path? Maybe we don't even need to change Dataplane model if we can do
```
  probes:
    endpoints:
    - inboundPath: /metrics
      inboundPort: 8081
      path: /8081/metrics
    - inboundPath: /pathToGrpcService # I don't remember syntax
      method: POST # In gRPC it's POST, right now AFAIR we only support GET
      inboundPort: 8082
      path: /8081/metrics
    port: 19000
```

Reply:
+1 to this I think some of my comments follow this point.

For TCPProbe @jakubdyszkiewicz seems right too:
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-tcp-liveness-probe

Though wasn't one of the point of these listeners to also bind sidecar probes with application ?

Reply:
TCP probes - true. I think TCP probes are very uncommon. Also by the time Envoy actually starts and receives config over XDS for inbound passthrough listener, TCP app next it is probably up already.

GRPC proves - true. But maybe we can route by host header? We would need to verify this, but when kubelet does gRPC HC to app on port 1234 and on port 1235, what does it send in host (:authority) header?

Reply:
I'd double check this assumption about the Host header. Usually http clients do the right thing of setting the host header to whatever is in the request (like Jakub suggests).
Let's be fact based please and not make assumptions.

Reply:
> An application running with other protocols would need TCP probes, for example, a gRPC application would need it before the cluster supports gRPC probes.

maybe... however gRPC probes went beta in 1.24 and GA with Kube 1.27 which is already EOL in two months (2024-06-28). What I've seen is that gRPC apps rather expose extra port with HTTP probes. We saw this pattern in our apps in Konnect as well.

Reply:
gRPC apps on legacy clusters is only an example. I think TCP probes are to support probing application that has other custom protocols.

Reply:
When gRPC requests are being sent to different ports, the HTTP hosts look like this: `10.42.0.12:5086`, which means the target port is part of the host header.

Reply:
For TCP probes, `tcp_proxy` listener with a  "blackhole" cluster will always pass the probe. So it looks like the only choice left is to transform TCP probes to HTTP probes.